In [ ]:
#FOR LOOP :-

In [2]:
#Q1
a1=range(1,11)
for i in a1:
    print(i,end=" ")

1 2 3 4 5 6 7 8 9 10 

Q2.
ans. A for loop is used when you know the number of iterations beforehand or when you want to iterate over a sequence (like a list or tuple).
A while loop is used when you want to execute a block of code repeatedly until a condition becomes False. It's suitable when you don't know the exact number of iterations in advance.







In [5]:
#Q3
total_sum=0
for i in range(1,101):
    total_sum+=i
print(f'sum of all no. from 1 to 100 is {total_sum}')



sum of all no. from 1 to 100 is 5050


Q4.
ans. In Python, you can iterate through a list using a for loop by iterating over each element in the list directly. Here's a simple example:

my_list = [1, 2, 3, 4, 5]
for element in my_list:
    print(element)

This loop iterates through each element in my_list, assigning each element to the variable element, and then performs some operation with it.








In [7]:
#Q5
numbers=[2,3,4,5,6,7,8,9]
product=1
for num in numbers:
    product*=num
print(f'The product of all elemrnt is {product}')

The product of all elemrnt is 362880


In [3]:
#Q6.
hii=list(filter(lambda a: a%2==0,range(1,21)))
print(hii)

[2, 4, 6, 8, 10, 12, 14, 16, 18, 20]


In [10]:
#Q7
test11=(lambda a: 1 if a ==1 else a*test11(a-1))

In [11]:
test11(5)

120

Q8.
ans.To iterate through the characters of a string using a for loop in Python, you can use the syntax for char in my_string. This loop will iterate over each character in the string my_string, assigning each character to the variable char one at a time.

In [24]:
#Q9.
a1=[1,2,4,56,8]
largest=a1[0]
for i in a1:
    if i > largest:
        largest=i

print(largest)


56


In [7]:
#Q10.
fib=int(input())
fibo=[0,1]
for i in range(2,fib+1):
    fib1=fibo[i-1]+fibo[i-2]
    fibo.append(fib1)
print(f'The number of fibo series {fibo[i]}')

The number of fibo series 5


In [5]:
#Q11
def find_vowel(user_input):
    vowel=['a','e','i','o','u']
    vowel_count=[]
    for char in user_input:
        if char.lower() in vowel:
            vowel_count.append(char)
    return vowel_count
user_input=input()
result=find_vowel(user_input)
print(f"The vowels in the string are:{result}")


The vowels in the string are:['a', 'i']


In [10]:
#Q12
for i in range(5):
    for m in range(i+1):
        print('*',end=' ')
    print()

* 
* * 
* * * 
* * * * 
* * * * * 


In [12]:
#Q13
list=['ankit','rahul']
for i in list:
    for m in i[::-1]:
        print(m,end=' ')

t i k n a l u h a r 

In [2]:
#Q14
def common_element(list1,list2):
    common=[]
    for element in list1:
        if element in list2 and element not in common:
            common.append(element)
    return common
list1=[1,2,3,4,5,6]
list2=[3,4,5,6,7,10]
result1=common_element(list1,list2)
print(f'Common elements between the lists:{result1}')

Common elements between the lists:[3, 4, 5, 6]


Q15.
ans- In Python, you can use a for loop to iterate through the keys and values of a dictionary using the .items() method. Here's how you do it:

my_dict = {'a': 1, 'b': 2, 'c': 3}

for key, value in my_dict.items():
    print(key, value)


This loop iterates through each key-value pair in the dictionary my_dict, assigning the key to the variable key and the corresponding value to the variable value, allowing you to access both key and value within the loop body.

In [3]:
#Q16
list3=int(input())
list4=int(input())
if list3>list4:
    smallest=list3
else:
    smallest=list4
    gcd=1
    for i in range(1,smallest+1):
        if list3%i==0 and list4%i==0:
            gcd=i
    print(f'gcd of this number :{gcd}')


gcd of this number :5


In [29]:
#Q17
input_string=input()
input22=input_string.lower()
is_palindrome=True
for char in range(len(input22)//2):
    if input22[char]!=input22[-char-1]:
        is_palindrome=False
        break
if is_palindrome:
    print(f'The input string is palindrome')
else:
    print(f"The input string is not palindrome")

The input string is palindrome


In [33]:
#Q18
entire_string=[1,1,2,2,3,3,4,5,6,7,7,8,8,9]
unique_list=[]
for item in entire_string:
    if item not in unique_list:
        unique_list.append(item)
print(f'The new list after removing duplicate is {unique_list}')


The new list after removing duplicate is [1, 2, 3, 4, 5, 6, 7, 8, 9]


In [34]:
#Q19.
entire_string=input()
counted_word=0
for words in str(entire_string):
    if words==' ':
        counted_word+=1
    else:
        counted_word+=1
print(f'The count of the word in a string is {counted_word}')

The count of the word in a string is 8


In [9]:
#Q20.
input12=int(input())
sum_of_num=0
for num in range(1,input12+1):
    if num%2!=0:
        sum_of_num+=num
print(f"The sum of all odd number from 1 to 50 is : {sum_of_num}")

The sum of all odd number from 1 to 50 is : 625


In [1]:
#Q21.
year=int(input())
leap_year=False
for num in range(1,year+1):
    if num%400==0 or num%4==0 and num%100!=0:
        if num==year:
            leap_year=True
            break
if leap_year:
    print(f'THe year is prime: {num}')
else:
    print(f"The year is not prime:{num}")

THe year is prime: 2024


In [22]:
#Q22.
square_root=int(input())
answer=[]
for i in range(2,square_root+1):
    root=i**0.5
    answer.append(root)
print(f'The roots of given number is : {answer}')


The roots of given number is : [1.4142135623730951]


In [23]:
#Q23.
def find_lcm(x, y):
    # Find the maximum of two numbers
    max_num = max(x, y)

    # Start checking from the maximum number
    for i in range(max_num, x * y + 1, max_num):
        if i % x == 0 and i % y == 0:
            return i

# Example usage:
num1 = 12
num2 = 15
print("LCM of", num1, "and", num2, "is:", find_lcm(num1, num2))



LCM of 12 and 15 is: 60


In [ ]:
# IF ELSE:

In [6]:
#Q1
num1=int(input())
if num>0:
    print(f"number is positive")
elif num<0:
    print(f"number is negative")
else:
    print(f'number is negative')

number is positive


In [7]:
#Q2
# Input a number
number = int(input("Enter a number: "))
if number%2==0:
    print(f"The given number is even number:{number}")
else:
    print(f"The given num is odd num :{number}")



The given number is even number:2


Q3.
ans- In Python, you can use nested if-else statements by placing one if-else statement inside another if-else block. Each if-else statement is evaluated sequentially, and the nested structure allows for more complex decision-making logic.


x = 10

if x > 5:

    print("x is greater than 5")
    if x > 7:
        print("x is also greater than 7")
    else:
        print("x is not greater than 7")
else:
    print("x is not greater than 5")


In this example, the outer if-else statement checks if x is greater than 5. If it is, it prints a message and then checks whether x is also greater than 7 using a nested if-else statement. Depending on the conditions, different messages are printed.


In [10]:
#Q4.
num1=int(input())
num2=int(input())
num3=int(input())
if num1>num2 and num1>num3:
    largest_num=num1
    print(f"largest num is : {largest_num}")
elif num2>num1 and num2>num3:
    largest_num=num2
    print(f"largest num is :{largest_num}")
else:
    largest_num=num3
    print(f'largest num is :{largest_num}')


largest num is :50


In [23]:
#Q5
num=int(input())
if num>0:
    print(f'The absolute value of entre num is :{num}')
else:
    print(f"The absolute value of entre num from negative is {-num} ")

The absolute value of entre num from negative is 20 


In [28]:
#Q6
entire_string=input()
vowel=['a','e','i','o','u','A','E','I','O','U']
is_constant=True
for char in entire_string:
    if char in vowel:
        is_constant=False
     
if is_constant:
    print(f'The entire string have constants')
else:
    print(f'The entire string have vowels')

The entire string have constants


In [32]:
#Q7
entre_age=int(input())
if entre_age>=18:
    print(f"you are eligible for vote")
else:
    print(f"you are not eligible for vote")

you are not eligible for vote


In [8]:
#Q8
entire_price=int(input())
discount=0
if entire_price<=1000:
    discount=0.15
elif entire_price<=500:
    discount=0.10
elif entire_price<=150:
    discount=0.05
else:
    print(f'Discount is not applicable on this price ' )

discount_rate=entire_price-(entire_price*discount)
print(f"The product price is :{entire_price:.2f}")
print(f"dicounted rate:{discount*100}%")
print(f"The payable amount after discount is :{discount_rate:.2f}")

Discount is not applicable on this price 
The product price is :5000.00
dicounted rate:0%
The payable amount after discount is :5000.00


In [2]:
#Q9.
lowest_limit=10
upper_limit=50
number=int(input())
if number>=lowest_limit and number<=upper_limit:
    print(f"the number {number} is in {lowest_limit} and {upper_limit}")
else:
    print(f"The number is not under the limit")

The number is not under the limit


In [9]:
#Q10
marks=int(input())
if marks>=95:
    grade='A++'
    print(f'you are pass your grade is :{grade}')
elif marks>=80:
    grade='A+'
    print(f'you are pass your grade is :{grade}')
elif marks>=70:
    grade='B'
    print(f'you are pass your grade is :{grade}')
elif marks>=50:
    grade='C'
    print(f'you are pass your grade is :{grade}')
else:
    print(f'your are failed')
    

you are pass your grade is :C


In [13]:
#Q11.
string=input()
if string== "":
    print(f"The string is empty")
else:
    print(f"THe string contains data")

THe string contains data


In [19]:
#Q12
angle1=int(input())
angle2=int(input())
angle3=int(input())
if angle1==angle2 and angle3:
    print(f"The triangle is equilaterial")
elif angle1==angle2 or angle2==angle3 or angle3==angle1:
    print(f"this triangle is isosceles")
else:
    print(f"This triangle is scalene")

The triangle is equilaterial


In [3]:
#Q13
day=int(input())
if day==1:
    week_day='monday'
    print(f"Today is :{week_day}")
elif day==2:
    week_day='tuesday'
    print(f"Today is :{week_day}")
elif day==3:
    week_day='wednesday'
    print(f"Today is :{week_day}")
elif day==4:
    week_day='thrusday'
    print(f"Today is :{week_day}")
elif day==5:
    week_day='friday'
    print(f"Today is :{week_day}")
elif day==6:
    week_day='saturaday'
    print(f"Today is :{week_day}")
elif day==7:
    week_day='sunday'
    print(f"Today is :{week_day}")
else:
    print(f"you entered the invalid input")

you entered the invalid input


In [6]:
#Q14
year=int(input())
if year%400==0 or year%4==0 and year%100!=0:
    print(f"THe entered year {year} is leap year")
else:
    print(f"THe  entered year {year} is not leap year")

THe  entered year 2023 is not leap year


Q15.
ans-In Python, the assert statement is used for debugging purposes to check whether a condition is true. If the condition evaluates to False, it raises an AssertionError exception. Here's how you can use it within if-else blocks:
for example:

x = 10

if x > 5:

    assert x > 5, "x should be greater than 5"  # Check if x is greater than 5
else:

    assert x <= 5, "x should be less than or equal to 5"  # Check if x is less than or equal to 5

print("End of program")



Replace condition, condition_is_true, and condition_else_is_true with your actual conditions. The optional message is printed if the condition fails. This helps in identifying issues during debugging.

In [11]:
#Q16
person_age=int(input())
if person_age>=60:
    print(f"Congratulations! You are eligible for a senior citizen discount. ")
else:
    print(f"Sorry, you are not eligible for a senior citizen discount.")

Congratulations! You are eligible for a senior citizen discount. 


In [16]:
#Q17.
char=input()
if char.isupper():
    print(f"The character '{char}' is uppercase")
elif char.islower():
    print(f"The char is '{char}' lowercase")
else:
    print(f"The char is neither lowercase or uppercase")

The char is 'ankit' lowercase


In [24]:
#Q19.
year=int(input())
if year%100==0:
    print(f"THe year {year} is century year")
else:
    print(f"The year {year} is not century year")

THe year 3000 is century year


In [29]:
#Q20.
angle1=int(input())
angle2=int(input())
angle3=int(input())
angle4=int(input())
print('size of angles is ',angle1,angle2,angle3,angle4)
if angle1==angle2 and angle3 and angle4:
    print(f"Its a square")
else:
    print(f"its not forming a proper square")

size of angles is  10 20 30 40
its not forming a proper square


Q21.
ans- continue: Skips the rest of the current iteration in a loop and moves to the next iteration.

break: Immediately exits the loop, skipping any remaining iterations.

These statements are used to control the flow of execution within loops based on certain conditions.

In [9]:
#Q22.
weight=float(input())
height=float(input())
height_cm=height/100
bmi=weight/(height_cm**2)
if bmi<18.5:
    category='Underweight'
elif 18.5<=bmi<24.5:
    category='Normalweight'
elif 24.5<=bmi<30:
    category='overweight'
else:
    category='obese'
print(f"your bmi is :{bmi:.2f}")
print(f'Your comes under category is :{category}')

your bmi is :24.15
Your comes under category is :Normalweight


Q23.
ans-
In Python, you can use the filter() function along with a lambda function containing an if-else statement to filter elements from a list. The filter() function applies a filtering function to each element of an iterable (like a list) and returns an iterator of the elements for which the filtering function returns True.

Here's a short example:

my_list = [1, 2, 3, 4, 5]

filtered_list = filter(lambda x: x % 2 == 0, my_list)  

print(list(filtered_list))  # Output: [2, 4]

In this example, lambda x: x % 2 == 0 is a lambda function that returns True if x is even, and False otherwise. The filter() function applies this lambda function to each element of my_list, keeping only those elements for which the lambda function returns True. Finally, list(filtered_list) converts the resulting iterator back into a list.

In [11]:
#Q24.
number=int(input())
if number>1:
    for i in range(2,int(number**0.5)+1):
        if number%i==0:
            print(f"The number {number} is not prime number")
            break
    else:
         print(f"The num {number} is prime number")
else:
    print(f'Not a valid input for prime check')


Not a valid input for prime check


In [ ]:
#MAP:-

Q1.
ans-
The map() function in Python is used to apply a given function to each item in an iterable (such as a list) and returns a new iterator that yields the results. It essentially applies the function to each element of the iterable, producing a new iterable with the results.

Here's a short example:


def square(x):
    return x ** 2


numbers = [1, 2, 3, 4, 5]
squared_numbers = map(square, numbers)


squared_numbers_list = list(squared_numbers)
print(squared_numbers_list)  # Output: [1, 4, 9, 16, 25]

In this example, the map() function applies the square() function to each element of the numbers list, producing a new iterable containing the squared values.

In [18]:
#Q2
number=[1,2,3,4,5,6,7,8]
list(map(lambda a: a*a,number))


[1, 4, 9, 16, 25, 36, 49, 64]

Q3.
ans-
The map() function and list comprehensions both offer ways to apply a function to every element in a sequence, like a list, but they differ in syntax and usage:

map() applies a function to each element of an iterable (like a list) and returns an iterator that yields the results.

List comprehensions provide a concise way to create lists by applying an expression to each item in an iterable.

Choosing between them depends on readability and personal preference:

map() can be more concise when applying a function to every element.

List comprehensions are often more readable and flexible, allowing for filtering and more complex expressions. Use them when readability is a priority or when additional complexity is needed.

In [25]:
#Q4.
nmae="ankit,kumar" ,'rahul'
set(map(lambda n:n.upper(),nmae))

{'ANKIT,KUMAR', 'RAHUL'}

In [27]:
#Q5
words=['ankit','rahul']
list(map(lambda a:len(a),words))

[5, 5]

Q6.
ans-
You can use the map() function to apply a custom function to elements of multiple lists simultaneously by passing the function as the first argument to map() and the lists as subsequent arguments. The function will be applied element-wise to corresponding elements of the lists.

In [2]:
#Q7.
celsius=[25,29,30,50,80,90]
list(map(lambda b:(b*9/5)+32,celsius))

[77.0, 84.2, 86.0, 122.0, 176.0, 194.0]

In [4]:
#Q8.
list22=[12.5,14.5,459.5,523.0]
list(map(lambda a :round(a),list22))

[12, 14, 460, 523]

In [ ]:
#REDUCE:-

Q1.
ans-
The reduce() function in Python is used to apply a function of two arguments cumulatively to the items of an iterable, from left to right, reducing the iterable to a single value. It's part of the functools module, so you need to import functools to use it.

Here's a basic example of reduce() usage to find the product of elements in a list:

from functools import reduce

numbers = [1, 2, 3, 4, 5]

product = reduce(lambda x, y: x * y, numbers)

print(product)  # Output: 120

in this example, reduce() applies the lambda function to each element of the list numbers, accumulating the result until a single value (product) is obtained.


In [2]:
#Q2
from functools import reduce
elements=[1,2,3,4,5]
reduce(lambda x,y : x*y ,elements)


120

In [4]:
#Q3
lisr12=[10,20,30,40,50]
reduce(lambda a,b:max(a,b),lisr12)

50

In [13]:
#Q4.
from functools import reduce
import operator

# Example list of strings
strings_list = ["hello", " ", "world"]

# Concatenate the list of strings into a single string
result = reduce(operator.add, strings_list)

print(result)  # Output: "hello world"


hello world


In [18]:
#Q5.
list55=int(input())
reduce(lambda x,y :x*y ,range(1,list55+1))

120

In [27]:
#Q6
import math
list66=[10,20,30,40]
reduce(math.gcd,list66)



10

In [34]:
#Q7
list00=[12,25,98,23,0]
reduce(lambda a,m :a+sum(int(digit) for digit in str(m) if digit.isdigit()),list00,0)

32

In [ ]:
#FILTER:-

Q1.
ans-The purpose of the filter() function in Python is to create a new iterator from elements of an iterable for which a specified function returns true. It returns an iterator that yields only the elements of the iterable for which the provided function returns True.

Example:

In [14]:
# Define a function to filter even numbers
def is_even(num):
    return num % 2 == 0

# Create a list of numbers
numbers = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]

# Use filter() to filter out even numbers
even_numbers = list(filter(is_even, numbers))

print(even_numbers)  # Output: [2, 4, 6, 8, 10]


[2, 4, 6, 8, 10]


In [35]:
#Q2.
list90=[1,2,3,4,5,8,9,1,0,12,6]
list(filter(lambda a:a%2==0,list90))

[2, 4, 8, 0, 12, 6]

In [36]:
#Q3.
name='ankit','rahul','pwskills','america'
list(filter(lambda a:a[0]=='a',name))

['ankit', 'america']

In [1]:
#Q4
list6=[1,2,3,4,5,6,78,9]
list(filter(lambda a:a%2!=0,list6))

[1, 3, 5, 9]

In [5]:
#Q6
woeds=['ankit','america','Bangolare']
list(filter(lambda a : len(a)<=6,woeds))

['ankit']

In [6]:
#q7
values=[10,20,30,40,50,60,70,80]
Threshold=39
list(filter(lambda v:v>Threshold,values))

[40, 50, 60, 70, 80]

In [ ]:
#RECURSION

Q1.
ans-
Recursion in Python is a programming technique where a function calls itself to solve smaller instances of the same problem. It involves breaking down a problem into smaller, more manageable subproblems until a base case is reached. Recursion is characterized by self-referential function calls.

Difference from iteration:

Recursion involves a function calling itself, while iteration involves looping constructs like for and while to repeatedly execute a block of code.
Recursion can be more concise and elegant for certain problems, but it can be less efficient in terms of memory usage and performance compared to iteration.
Recursion typically involves more overhead due to function call overhead and maintaining a call stack.

In [5]:
#Q2
def fact(n):
    if n<=1:
        return 1
    else:
        return n*fact(n-1)

In [6]:
fact(5)

120

In [9]:
#Q3
def fibo(q):
    if q<=1:
        return q
    else:
        return fibo(q-2)+fibo(q-1)

In [10]:
fibo(5)

5

In [5]:
#Q4
def calculate(*args):
    if not args:
        return 0
    else:
        return args[0]+calculate(*args[1:])

In [8]:
calculate(1,2,3,4)

10

Q5.
ANS-o prevent a recursive function from running indefinitely and causing a stack overflow error, you can:

Ensure that the base case(s) are properly defined to terminate the recursion.

Make sure that the inputs are getting closer to the base case(s) in each recursive call.

Limit the depth of recursion or impose a maximum recursion depth if necessary.

Use iterative solutions or tail recursion where applicable to avoid excessive stack usage.

In [ ]:
#Q6.

In [9]:
#Q7.
def reverse(l):
    if l=="":
        return ""
    else:
        return reverse(l[1:])+l[0]

In [10]:
reverse('ankit')

'tikna'

In [17]:
#Q8.
def power(n,l):
    if l==0:
        return 1
    else:
        return n*power(n,l-1)

In [18]:
power(2,5)

32

In [19]:
#Q9.


Permutations of abc : ['abc', 'acb', 'bac', 'bca', 'cab', 'cba']


In [22]:
#Q10.
def palindrome(j):
    if j=="":
        return True
    elif j==j[::-1]:
        return True
    else:
        return False

In [23]:
palindrome('ankit')

False

In [4]:
#Q11
def generate_combinations(elements, combo_length):
    if combo_length == 0:
        return [[]]
    if len(elements) == 0:
        return []
    first_element = elements[0]
    rest_of_elements = elements[1:]
    combinations_without_first = generate_combinations(rest_of_elements, combo_length)
    combinations_with_first = generate_combinations(rest_of_elements, combo_length - 1)
    for combo in combinations_with_first:
        combo.append(first_element)
    return combinations_without_first + combinations_with_first

# Example usage:
elements_list = [1, 2, 3, 4]
combo_length = 3
print(generate_combinations(elements_list, combo_length))


[[4, 3, 2], [4, 3, 1], [4, 2, 1], [3, 2, 1]]


BASICS OF FUNCTIONS

Q1.
ANS-A function in Python is a block of reusable code that performs a specific task. It takes input (if any), performs operations, and returns output (if specified). Functions are used to organize code, make it more readable, and reduce repetition by allowing you to encapsulate functionality into modular units. They promote code reuse, improve maintainability, and make programs easier to understand and debug.

Q2.
ANS-In Python, you define a function using the def keyword followed by the function name and parentheses containing any parameters. Here's a simple example:

In [9]:
def greet(name):
    print("Hello, " + name + "!")

# Example usage:
greet("Alice")


Hello, Alice!


Q3.
ANS-A function definition defines what a function does, including its name, parameters, and the code it executes. It's like creating a recipe for a dish.

A function call is when you actually use the function by its name and provide any necessary arguments. It's like following the recipe and making the dish.







In [10]:
#Q4.
def calculate_sum(num1, num2):
    return num1 + num2

result = calculate_sum(3, 5)
print("The sum of the two numbers is:", result)


The sum of the two numbers is: 8


Q5.
Ans A function signature is a way to describe a function using its name and the types of its parameters and return value. It typically includes the function name, the types of parameters it accepts (if any), and the type of value it returns (if any). It does not include the function's implementation or body.

In [12]:
#Q6
def product(x, y):
    return x * y

result = product(3, 5)
print(result)  



15


Function Parameters and Arguments:


Q1
Ans-Formal parameters are the placeholders in the function definition that represent the data that the function expects to receive when it's called.
Actual arguments (or arguments) are the actual values or variables passed to the function when it's called, which correspond to the formal parameters.
In simpler terms, formal parameters are like empty slots waiting to be filled with data, and actual arguments are the actual pieces of data that fill those slots when the function is called.

In [13]:
#Q2
# Define a function with default argument values
def greet(name="World"):
    print("Hello,", name)

# Call the function without providing any argument
greet()

# Call the function providing a specific argument
greet("Alice")


Hello, World
Hello, Alice


Q3.
Ans-in Python, you use keyword arguments by specifying the parameter name followed by the value you want to pass to that parameter, separated by an equal sign =. Here's a simple example:

In [14]:
def greet(name, message):
    print(f"Hello, {name}! {message}")

# Using keyword arguments
greet(name="Alice", message="How are you?")


Hello, Alice! How are you?


In [1]:
#Q4.
def calculate_sum(*args):
    return sum(args)

# Example usage:
result = calculate_sum(1, 2, 3, 4, 5)
print(result) 


15


Q5.
ANS-In short, the *args syntax allows a function to accept any number of positional arguments, while the **kwargs syntax allows it to accept any number of keyword arguments in the form of a dictionary. They provide flexibility when defining functions without specifying the exact number of arguments they'll receive.






Return Values and Scoping:


Q1.
Ans-The return statement in Python functions is used to specify the value that the function should give back when it's called. It stops the execution of the function and sends the specified value back to the caller.

In [3]:
def divide_and_remainder(dividend, divisor):
    quotient = dividend // divisor
    remainder = dividend % divisor
    return quotient, remainder

   # This will return (3, 1)


In [4]:
divide_and_remainder(10, 3)

(3, 1)

Q2.
Ans-In Python, variable scope refers to where in your code a variable is accessible and can be used.

Local variables are defined within a function and are accessible only within that function. They cannot be accessed from outside the function.

Global variables are defined outside of any function and can be accessed from anywhere in the code, including inside functions. However, if you want to modify a global variable within a function, you need to explicitly declare it as global inside the function.

In short, local variables are limited to the function they're defined in, while global variables can be accessed and modified throughout the entire program.

In [4]:
#Q3.
# Define a global variable
global_var = 10

# Define a function that uses the global variable
def increment_global():
    global global_var  # Indicate that we want to modify the global variable
    global_var += 1

# Define a function that prints the global variable
def print_global():
    print("Global variable:", global_var)

# Demonstrate the use of global variable within functions
print_global()    # Output: Global variable: 10
increment_global()
print_global()    # Output: Global variable: 11


Global variable: 10
Global variable: 11


In [5]:
#Q4.
def fact(a):
    if a <=0:
        return 1
    else:
        return a*fact(a-1)


In [6]:
fact(5)

120

Q5.
Ans-In Python, you can access variables defined outside a function from within the function by directly referencing them using their names. These variables are considered to be in the function's "scope" and are accessible as long as they are not shadowed by variables with the same name within the function.

Lambda Functions and Higher-Order Functions:



Q1.
Ans-Lambda functions in Python are small, anonymous functions defined using the lambda keyword. They are typically used when you need a simple function for a short period of time, often in situations where you would use a regular function but don't want to bother with the formalities of defining one.

Lambda functions are commonly used in scenarios such as:

Passing a simple function as an argument to higher-order functions like map(), filter(), and sorted().
Defining small, inline functions within list comprehensions or dictionary comprehensions.
Creating short, concise callbacks for event handlers or callback functions.
Overall, lambda functions offer a convenient way to create quick, one-time functions without the need for a formal function definition.








In [9]:
#Q2.
# List of tuples
list_of_tuples = [(1, 3), (10, 1), (33, 2)]

# Sorting the list of tuples based on the second element
sorted_list = sorted(list_of_tuples, key=lambda x: x[1])

# Printing the sorted list
print(sorted_list)



[(10, 1), (33, 2), (1, 3)]


In [10]:
#Q3.
def apply_operation(operation, x, y):
    return operation(x, y)

def add(x, y):
    return x + y

def multiply(x, y):
    return x * y

result1 = apply_operation(add, 3, 4)  # Passing the add function as an argument
result2 = apply_operation(multiply, 3, 4)  # Passing the multiply function as an argument

print(result1)  
print(result2)


7
12


In [11]:
#Q4.
def apply_function_to_list(numbers, func):
    return [func(num) for num in numbers]

# Example usage:
numbers_list = [1, 2, 3, 4, 5]
square_function = lambda x: x ** 2
result = apply_function_to_list(numbers_list, square_function)
print(result)


[1, 4, 9, 16, 25]


Built-in Functions:


Q1.
ans-len(): Returns the length of an object, such as a list, tuple, string, etc. It counts the number of elements in the object.

max(): Returns the maximum value from a sequence, such as a list, tuple, or set. It determines the largest element in the sequence.

min(): Returns the minimum value from a sequence, such as a list, tuple, or set. It determines the smallest element in the sequence.

These functions are handy for various tasks such as finding the size of a collection, determining extreme values, or simply comparing elements.

In [14]:
#Q2.
# Define a function to be applied to each element
def square(x):
    return x ** 2

# Create a list of numbers
numbers = [1, 2, 3, 4, 5]

# Use map() to apply the function to each element of the list
squared_numbers = list(map(square, numbers))

# Print the squared numbers
print(squared_numbers)


[1, 4, 9, 16, 25]


Q3.
ans-The filter() function in Python is used to create a new iterator from elements of an iterable (like a list) for which a provided function returns true. It takes two parameters: the function that determines the filtering condition and the iterable to be filtered.

You would use filter() when you need to selectively pick elements from an iterable based on certain criteria defined by a function, rather than iterating over the entire list manually and checking each element individually. It's particularly useful for compactly expressing filtering logic.

In [15]:
#Q4.
from functools import reduce

# Define a list of numbers
numbers = [2, 3, 4, 5]

# Use reduce() function to find the product of all elements
product = reduce(lambda x, y: x * y, numbers)

# Print the product
print("Product of all elements:", product)


Product of all elements: 120


Function Documentation and Best Practices:

Q1.
ans-Docstrings in Python functions serve the purpose of providing documentation or information about the function's purpose, parameters, and return values. They help developers understand how to use the function without having to inspect its implementation.

To write a docstring, you enclose a multi-line string within triple quotes (''' or """) right after the function definition. It's typically formatted like:

In [24]:
#example:
def function_name(parameters):
    """
    Description of the function.

    Parameters:
    - parameter1: Description of parameter1
    - parameter2: Description of parameter2

    Returns:
    Description of the return value.
    """
    # Function implementation


Q2.
ans- 
1. Use descriptive names that clearly indicate the purpose or functionality of the function or variable.

2. Follow the snake_case naming convention for functions and variables, where words are separated by underscores.

3. Use lowercase letters for function and variable names, except for class names which typically use CamelCase.
4. Choose meaningful and concise names, avoiding overly long or cryptic names.
5. Be consistent with naming conventions throughout your codebase.
6. Avoid using reserved keywords or built-in function names as identifiers to prevent conflicts and confusion.
7. Use nouns for variables and verbs for functions to convey their roles and actions.
8. Use singular nouns for variables representing single items and plural nouns for collections or sequences of items.
9. Consider using prefixes or suffixes to indicate the type of variable (e.g., list_of_names, num_elements).
10. Refactor names when necessary to improve clarity and maintainability, especially as the codebase evolves.